refactor: Pokemon IDs are now unique and can't be `0`

DayKev opened the pull request:
## What are the changes the user will see?
Less random bugs due to multiple Pokémon having the same ID. This primarily affects the Twins trainer battle.

## Why am I making these changes?
If multiple Pokémon have the same ID (or an ID less than `4`), things break.

## What are the changes from a developer perspective?
- `BattleScene#nextPokemonID` added, which is the next ID that will be assigned to a Pokémon. This is a simple incrementing integer.
- `BattleScene#getNextPokemonID()` added to get the next ID to assign to a Pokémon and then increment `nextPokemonID`.
- `BattleScene#updateNextPokemonID()` added which sets `nextPokemonID` to the max of the passed in value and the current value of `nextPokemonID`. Used when loading from a save to prevent ID collisions from pre-existing pokemon.
- The `Pokemon` constructor calls `BattleScene#getNextPokemonID()` to assign an ID when not loading from a save.
- `Pokemon#personalityValue` added which is a non-unique random integer from `0` to `2^32 - 1` that is used to [determine shininess](https://bulbapedia.bulbagarden.net/wiki/Personality_value#Shininess) instead of the Pokémon's ID.

## Checklist
- [x] Otherwise: **I'm using `beta` as my base branch**
- [x] There is no overlap with another PR?
- [x] The PR is self-contained and cannot be split into smaller PRs?
- [x] Have I provided a clear explanation of the changes?
- [x] Have I tested the changes manually?

DayKev committed:
[Refactor] Pokemon IDs are now unique and can't be `0`

DayKev committed:
Make `Pokemon#generateId()` return the generated ID

This allows mocking the implementation of the method
in `RngHelper#equalSample()` so that the infinite loop
caused by mocking `Phaser.Math.RND.realInRange`
can be avoided

DayKev committed:
`Pokemon#generateId` no longer needs to be `protected`

DayKev committed:
Merge branch 'beta' into refactor/make-pokemon-ids-unique

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -500,11 +502,22 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+      return this.id;
+    }
+
+    let id = randSeedInt(4294967295, 1);
+    while (globalScene.activePokemonIDs.has(id)) {
+      id = randSeedInt(4294967295, 1);
+    }
```
Does this need to be randomly generated if IDs will be checked from a set? Theoretically you could just take the largest number from the set and just add 1 to it or something if a unique identifier is required

DayKev replied:
True, but it was simpler to do this than add handling to make sure the number doesn't go above `2 ** 32 - 1`.

Tempo-anon replied:
If the only thing we care about is ensuring that there are no collisions then this could just be some kind of iterator especially since there's a upper limit to how many Pokemon will even exist in the course of a run. (Quick napkin math example of 6 * 6000 = 36000)

DayKev replied:
Wdym?

Tempo-anon replied:
I'm just wondering about the usecase of ID. From my understanding it's just something to uniquely identify a Pokemon in the context of a specific run. Since this PR updates the global scene to keep track of all existing IDs I'm wondering if instead of a set of numbers that still get randomly generated the global scene is now an int that starts with 1 and every new Pokemon generated is set equal to that number and then increments it.

in battle-scene.ts
```
public nextPokemonId = 1;
```

in pokemon.ts
```
this.id = globalScene.nextPokemonId;
globalScene.nextPokemonId += 1;
```

Note that this means that when a Pokemon is destroyed the `nextPokemonId` is not decremented but that shouldn't matter as we're only keeping track of information by the int

MokaStitcher replied:
It's a flair thing, similar to mainline. And especially if we intend to implement multiplayer stuff i feel like this solution would add even more headache than

Tempo-anon replied:
Well that's what I'm trying to clarify here. If the only use of the `id` field in the `Pokemon` class is to distinguish individual Pokemon then we can just make it an int and have the global scene keep track of the next id to give out. In terms of multiplayer I imagine there's still only going to be one global scene and the whole process of getting and incrementing the id being tracked in the global scene to be atomized.

The personality value of a Pokemon in the mainline games is also used to calculate other values like shininess and gender/ability/nature/characteristic/forms/size/evolution in addition to identifying a Pokemon.

DayKev replied:
We use Pokémon IDs to calculate shininess matching the mainline method.
I think they might be used for one other thing like that, not sure.

If we weren't using them for anything other than identifiers then yeah I would've just made them unique incrementing numbers or something.

Tempo-anon replied:
Hm okay.

Very very minor but this means the chances of shinies will change based on how many IDs are in the set of existing IDs at a certain time.

MokaStitcher replied:
Doesn't seem minor to me at all, this number could be manipulated. Since eggs will contribute to it, you could get any guaranteed shiny from same species eggs if you know what you are doing
If a shiny id exists in the very low numbers it would also be noticed immediately by players
And if there is none... Well players won't get shinies unless they play for an extended period of time

MokaStitcher replied:
Or do you mean with the current implementation from this PR? In which case I'm not sure I follow. The chance of pulling 2 exact same IDs naturally is almost non existent. If it ever happens it's because we messed something up (or the twins)
Which makes me wonder if we should print a console warning or something whenever an id gets rerolled

Tempo-anon replied:
I was just asking if the id was being used for anything at all or if it was just an identifier. If it is just an identifier then I think my suggestion of just keeping a counter that increments would make the most sense but apparently it is being used 🤷 

Though now that I think about it I believe the shiny check is only done when a new Pokemon is created. Starting a run with 6 shinies won't affect new shiny rates at all but catching a shiny and then keeping it on the team will (probably a negligible amount)

Eggs hatching new Pokemon with their own IDs..... I'm not as familiar with the exact technicals of that but I wouldn't be surprised if it creates an actual `Pokemon` object that then needs its own ID that's also being used for shiny checking.

You can probably ignore most of my rambling. I don't think it matters too much but might be worth documenting somewhere.

MokaStitcher replied:
> I wouldn't be surprised if it creates an actual Pokemon object that then needs its own ID that's also being used for shiny checkin

They do

DayKev replied:
Eggs do create `PlayerPokemon` objects, though I believe they're properly deleted afterwards now.

In mainline, Pokémon PVs aren't unique afaik, so I could add a new PV field to the `Pokemon` class and then change it so the things that are using Pokémon IDs as PVs here instead use the new PV field (which no longer needs to be unique). Then make the existing ID field just an incrementing (but unique) number.

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -500,11 +502,22 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
+      return this.id;
+    }
+
+    let id = randSeedInt(4294967295, 1);
```
The collision rate of 2^32 is 50% with 2^16 trials. This seems pretty unlikely in any run

DayKev replied:
In any individual run it probably won't happen, but over the lifetime of the game and over every run played, it will probably happen.

Tempo-anon commented on src/field/pokemon.ts:
```diff
@@ -500,11 +502,22 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
   /**
-   * Sets this Pokemon's ID to be a random integer from 0 to 2^32 - 1, inclusive.
-   * @todo This should be `protected` or `private` but MEs currently call it
+   * Generates a random integer from `1` to `2^32 - 1`, inclusive, to be assigned to as the Pokemon's ID.
```
What's the special issue with the ID being 0?

DayKev replied:
There's a bunch of places where things like arena/battler tags have a `sourceId` field that can be `undefined`, and the checks to see if it's `undefined` are `if (this.sourceId)` which also wouldn't pass if the `sourceId` was `0`. It's simpler to just prevent IDs of `0` than change every single `if` statement lol.

Tempo-anon reviewed (commented):
Does this fix the issue with twins having Pokemon with the same ID and causing bugs to happen? :pikasmile:

DayKev commented:
> Does this fix the issue with twins having Pokemon with the same ID and causing bugs to happen? :pikasmile:

The Twins' Pokémon have different IDs (dunno if it was this PR that fixed it, I didn't check if they had different IDs on beta).

MokaStitcher commented:
> > Does this fix the issue with twins having Pokemon with the same ID and causing bugs to happen? :pikasmile:
> 
> The Twins' Pokémon have different IDs (dunno if it was this PR that fixed it, I didn't check if they had different IDs on beta).

They had the same Id before, because they are generated with the same seed on purpose so that their mons match (plusle with minun for instance). Which is why I asked you to check how twins behaved with your changes. They seem to no longer have the same if buy retain the matching mon thing, which is good

MokaStitcher commented:
See #598

DayKev committed:
Change `id` to be a simple incrementing number

Add `Pokemon#personalityValue` to be used to determine shininess;
this is not a unique value

Remove obsolete `Pokemon#generateId` mock from `RngHelper#equalSample`

DayKev committed:
Update TSDoc for `Pokemon#generateId`

DayKev committed:
Update TSDoc for `Pokemon#trySetShiny`

DayKev committed:
Use `source.id` for old saves instead of `0`

DayKev committed:
Merge branch 'beta' into refactor/make-pokemon-ids-unique

DayKev commented:
Updated to use basic incrementing integers for the Pokemon's ID. Shininess now uses `Pokemon#personalityValue` (which is a non-unique integer from `0` to `2^32 - 1`) instead of `Pokemon#id`.

MokaStitcher commented on src/field/pokemon.ts:
```diff
@@ -512,9 +519,9 @@ export abstract class Pokemon extends Phaser.GameObjects.Container {
     }
 
-    let id = randSeedInt(4294967295, 1);
+    let id = 1;
```
If we're going with this I think we might as well defer that logic to the scene itself
Call `globalScene.getNewPokemonID()` or something?

MokaStitcher replied:
This way we don't make the activeIDs list public and can use helper functions to add/remove them instead

DayKev replied:
True.

DayKev committed:
Move Pokemon ID handling to `BattleScene`, make `Pokemon#id` `readonly`

DayKev committed:
Merge branch 'beta' into refactor/make-pokemon-ids-unique

DayKev committed:
Merge branch 'beta' into refactor/make-pokemon-ids-unique

DayKev committed:
Merge branch 'beta' into refactor/make-pokemon-ids-unique

SirzBenjie reviewed (approved):
This is a solid way of handling it.

MokaStitcher commented:
Does this requires server side updates due to new pv field?

MokaStitcher commented on src/battle-scene.ts:
```diff
@@ -725,6 +730,19 @@ export class BattleScene extends SceneBase {
+  }
+
+  /**
+   * Sets {@linkcode nextPokemonID} to the maximum of its current value and the passed in value plus 1. \
```
for my own curiosity, what does the `\ ` do?

DayKev replied:
A `\` at the end of a line in Markdown creates a linebreak, and TSDocs use Markdown.

DayKev replied:
Otherwise you need to have a blank line in-between two lines to create a linebreak in Markdown.

MokaStitcher replied:
Ah I see, thanks 👍

MokaStitcher reviewed (dismissed):
lgtm

We may want to consider resetting nextPokemonID to 1 in `scene.reset()` if we want to avoid the ids in a new run to sort of depend on the ids in previous run.
It shouldn't be an issue most of the time but if we imagine that a player loads an existing save, then finishes the run and starts a new one, the ids in that new run would depend on the IDs in the first run. Now if we assume that every time they load the game they load an existing run and later start another run, the IDs would keep snowballing up and up over time, leading to things imploding eventually. It's quite an unlikely scenario because we'd need to reach a very high number, but considering that at the moment Pokemon objects are being created quite liberally in different parts of the code, it's probably better to be safe than sorry :')
I think having a reset point for the ids should help prevent this risk

DayKev commented:
> We may want to consider resetting nextPokemonID to 1 in `scene.reset()` if we want to avoid the ids in a new run to sort of depend on the ids in previous run.

Oh yeah, I forgot to do that.

DayKev committed:
Merge branch 'beta' into refactor/make-pokemon-ids-unique

DayKev committed:
Reset `BattleScene#nextPokemonID` in `BattleScene#reset()`

DayKev committed:
Fix outdated comment

DayKev committed:
`BattleScene#nextPokemonID` now starts at 4
